expirationd.lua

```diff
@@ -100,13 +102,19 @@ local function default_do_worker_iteration(task)
     local space_len = task.vinyl_assumed_space_len
     local checked_tuples_count = 0
     local vinyl_checked_tuples_count = 0
+    if task.atomic_iteration then
```

Comment:
Will someone close transaction if task:iterate_with/expiration_process(task, tuple)/... throws an error?

Reply:
I think you can even try to reproduce it for vinyl. Expiration task delete some values in transaction another code updates the same values. Finally you transaction will be rolled back because of conflict.

Probably it's really hard to reproduce but I hope it's possible if you will give few memory for vinyl - it lead to more often yields because it will read and write directly to the disk.

Reply:
I think these tests should be in the tarantool / tarantool repository, in my opinion, this is not entirely relevant for this task

Reply:
No. It's responsibility of user code to close transactions. Your code could turn Tarantool in inconsistent condition. Please fix it and handle all errors properly.

Reply:
After transaction throws an error fiber will die.

Should user manually restart it's task after your unsafe code terminate this task? Even task will be alive you can't open transaction again. Because box.begin() could be called twice.

```
tarantool> box.begin() box.begin()
---
- error: 'Operation is not permitted when there is an active transaction '
...
```

Do not justify the unsafety of your code by saying that Tarantool should fix all this for you.

Reply:
After personal messaging, we understood that the error will fall into pcall and the fiber will die and a new one will start. For exact rollback of the transaction, box.rollback() was added before the fiber dies

Worker loop:
```lua
...
local state, err = pcall(task.do_worker_iteration, task)
if state then
    task.on_full_scan_success(task)
else
    task.on_full_scan_error(task, err)
end

task.on_full_scan_complete(task)
if not state then
    box.rollback()
    error(err)
end
...
```

```lua
local function get_fiber_id(fiber)
    local fid = 0
    if fiber ~= nil and fiber:status() ~= "dead" then
        fid = fiber:id()
    end
    return fid
end

local function guardian_loop(task)
    -- detach the guardian from the creator and attach it to sched
    fiber.name(string.format("guardian of %q", task.name), { truncate = true })

    while true do
        -- if fiber doesn't exist
        if get_fiber_id(task.worker_fiber) == 0 then
            -- create worker fiber
            task.worker_fiber = fiber.create(worker_loop, task)

            log.info("expiration: task %q restarted", task.name)
            task.restarts = task.restarts + 1
        end
        fiber.sleep(constants.check_interval)
    end
end
```